app/scripts/modules/core/src/managed/constraints/registry.tsx

```diff
@@ -0,0 +1,175 @@
+import { DateTime } from 'luxon';
+import React from 'react';
+
+import { IconNames } from '@spinnaker/presentation';
+
+import { DeploymentWindow } from '../artifactDetail/constraints/DeploymentWindow';
+import {
+  ConstraintStatus,
+  IAllowedTimesConstraint,
+  IBaseConstraint,
+  IConstraint,
+  IDependsOnConstraint,
+  IManagedArtifactVersionEnvironment,
+} from '../../domain';
+import { BasePluginManager } from '../plugins/BasePluginManager';
+
+const NO_FAILURE_MESSAGE = 'no details available';
+const UNKNOWN_CONSTRAINT_ICON = 'mdConstraintGeneric';
+
+const constraintHasNotStarted: ConstraintStatus[] = ['PENDING', 'NOT_EVALUATED'];
+interface IConstraintOverrideAction {
+  title: string;
+  pass: boolean;
+}
+
+export const hasSkippedConstraint = (constraint: IConstraint, environment: IManagedArtifactVersionEnvironment) =>
+  environment.state === 'skipped' && constraintHasNotStarted.includes(constraint.status);
+export interface IConstraintHandler<K = string> {
+  /** The type of the constraint - versioning is supported by adding @{version}, e.g. myConstraint@ */
+  kind: K;
+  /** The icon can be a string (from IconNames) or a partial map from statuses to IconNames */
+  iconName: IconNames | { [status in ConstraintStatus | 'DEFAULT']?: IconNames };
+
+  /** Render function of the constraint title */
+  titleRender: (constraint: IBaseConstraint | IConstraint) => React.ReactNode;
+
+  /** Optional render function of the constraint description */
+  descriptionRender?: (constraint: IBaseConstraint | IConstraint) => React.ReactNode;
```

Comment:
Wondering if this API could be typed as a `ComponentType<IConstraint>` instead?  I'm personally more comfortable working with components rather than passing JSX  around.  

Just a thought, not a strong opinion

Reply:
Sure, sounds good to me, and then change the render function to this? Is that what you were thinking of?

```
  renderTitle(constraint: IConstraint): React.ReactNode {
    const Component = this.getHandler(constraint.type)?.titleRender;
    if (Component) {
      return <Component constraint={constraint} />
    }
    return `${constraint.type} constraint - ${constraint.status}`;
  }
```

Reply:
Yeah, it cleans up the registry file so I'm good with it :)